docker/credentials_helper.go

```diff
@@ -72,18 +73,47 @@ func GetCredentialsFromCredsStore(registry string) (types.DockerAuthConfig, erro
 	return result, fmt.Errorf("failed to get credentials from helper specified in ~/.docker/config.json: %w", ErrCredentialsNotFound)
 }
 
-func to2ndLevelDomain(rawurl string) string {
-	if !strings.Contains(rawurl, "://") {
-		rawurl = "https://" + rawurl
+func hostPort(registry string) (host string, port string) {
+	host, port = registry, "443"
+	if !strings.Contains(registry, "://") {
+		h, p, err := net.SplitHostPort(registry)
+
+		if err == nil {
+			host, port = h, p
+			return
+		}
+		registry = "https://" + registry
 	}
-	u, err := url.Parse(rawurl)
+
+	u, err := url.Parse(registry)
 	if err != nil {
-		return ""
+		panic(err)
 	}
-	hostname := u.Hostname()
-	parts := strings.Split(hostname, ".")
-	if len(parts) <= 1 {
-		return hostname
+	host = u.Hostname()
+	if u.Port() != "" {
+		port = u.Port()
+	} else if u.Scheme == "http" {
+		port = "80"
 	}
-	return parts[len(parts)-2] + "." + parts[len(parts)-1]
+	return
+}
+
+// checks whether registry matches in host and port
+// with exception where port 80 and 443 are considered equal
```

Comment:
I don't really understand why the two different ports are considered equal.

Reply:
I also not 100% sure about this. My idea here is that if you are running registry on `80` and also `443` then most likely the two sockets are serving the same registry. But if you run registry on `5000` and `5001` then most likely the two socket are serving two distinct registries.

Reply:
but maybe it would make more sense to make `80` and `443` equivalent not to each other but to empty port `""`.

Reply:
> but maybe it would make more sense to make `80` and `443` equivalent not to each other but to empty port `""`.

This makes a bit more sense to me.